lib/src/editor.dart

```diff
@@ -0,0 +1,790 @@
+  /// error, but `null` (as opposed to [YamlScalar] null) will be returned if the operation would
+  /// have resulted in a `null` value on a dart collection.
+  ///
+  /// ```dart
+  /// final doc = YamlEditor('{a: {d: 4}, c: ~}');
+  /// doc.parseValueAt(['b', 'd']); // ArgumentError
+  /// doc.parseValueAt(['b']); // null
+  /// doc.parseValueAt(['c']); // YamlScalar(null)
+  ///
+  /// final doc2 = YamlEditor('[0,1]');
+  /// doc2.parseValueAt([2]); // ArgumentError
+  /// doc2.parseValueAt(["2"]); // ArgumentError
+  /// ```
+  YamlNode parseValueAt(Iterable<Object> path) {
+    dynamic current = _contents;
+
+    for (var key in path) {
+      try {
+        current = current.nodes[key];
+      } catch (Error) {
+        throw ArgumentError(
+            'Invalid path $path: Invalid key $key supplied to $current');
+      }
+    }
+
+    return current;
+  }
+
+  /// Sets [value] in the [path]. Takes an optional [style] parameter.
+  ///
+  /// If the [path] is not accessible (e.g. it currently does not exist in the document),
+  /// an error will be thrown. Note that [setIn] provides a different result as compared to
+  /// a [removeIn] followed by an [insertIn], because it preserves comments at the same level.
+  ///
+  /// ```dart
+  /// final doc = YamlEditor('''
+  ///   - 0
+  ///   - 1 # comment
+  ///   - 2
+  /// ''');
+  /// doc.setIn([1], 'test');
+  /// ```
+  ///
+  /// Expected Output:
+  /// '''
+  ///   - 0
+  ///   - test # comment
+  ///   - 2
+  /// '''
+  ///
+  /// ```dart
+  /// final doc2 = YamlEditor("[YAML Ain't Markup Language   # comment]");
+  /// doc2.removeIn([1]);
+  /// doc2.insertInList([1], 'test');
+  /// ```
+  ///
+  /// Expected Output:
+  /// '''
+  ///   - 0
+  ///   - test
+  ///   - 2
+  /// '''
+  void setIn(Iterable<Object> path, Object value, {YamlStyle style}) {
+    final collectionPath = path.take(path.length - 1);
+    final yamlCollection = parseValueAt(collectionPath);
+    final lastNode = path.last;
+
+    var edit;
+    var expectedNode;
+
+    final valueNode = _yamlNodeFrom(value);
+
+    style ??= defaultStyle;
+
+    if (yamlCollection is YamlList) {
+      edit = _setInList(_yaml, yamlCollection, lastNode, value, style);
+      expectedNode = _updatedYamlList(
+          yamlCollection, (nodes) => nodes[lastNode] = valueNode);
+    } else if (yamlCollection is YamlMap) {
+      edit = _setInMap(_yaml, yamlCollection, lastNode, value, style);
+      final keyNode = _yamlNodeFrom(lastNode);
+      expectedNode = _updatedYamlMap(
+          yamlCollection, (nodes) => nodes[keyNode] = valueNode);
+    } else {
+      throw TypeError();
+    }
+
+    _performEdit(edit, collectionPath, expectedNode);
+  }
+
+  /// Appends [value] into the list at [listPath], only if the element at the given path
+  /// is a [YamlList]. Takes an optional [style] parameter.
+  void addInList(Iterable<Object> listPath, Object value,
```

Comment:
I somehow feel iffy about the untyped `path`.

One could consider taking a YamlNode(from the original document) and use that as the marker of where to do the edit:
```
void setIn(YamlList list, Object value);
```

That would require YamlEditor to give access to the parsed YamlDocument.

Not sure it is much better in practice.

Reply:
I think we would have to give up the `YamlEditor` then, because as soon as you make one change then the previously retrieved `YamlNode`s are invalidated...